autode/wrappers/QChem.py

```diff
@@ -0,0 +1,502 @@
+                # e.g.  Q-Chem 5.4.1 for Intel X86 EM64T Linux
+                str0, str1 = line.split()[:2]
+
+                if str0 == 'Q-Chem' and '.' in str1 and ',' not in str1:
+                    return str1
+
+        return '???'
+
+    def execute(self, calc) -> None:
+        """Execute a qchem calculation"""
+
+        @work_in_tmp_dir(filenames_to_copy=calc.input.filenames,
+                         kept_file_exts=('.in', '.out'))
+        def execute_qchem():
+            params = [self.path, '-nt', str(calc.n_cores), calc.input.filename]
+            run_external(params, output_filename=calc.output.filename)
+
+        execute_qchem()
+        return None
+
+    def calculation_terminated_normally(self, calc) -> bool:
+        """Did the calculation terminate normally?"""
+
+        if not calc.output.exists:
+            logger.warning('QChem output did not exist - calculation '
+                           'did not terminate')
+            return False
+
+        calc_started = False
+
+        for line in calc.output.file_lines:
+            if 'Q-Chem begins' in line:
+                calc_started = True
+
+            if 'MAXIMUM OPTIMIZATION CYCLES REACHED' in line:
+                logger.info('Maximum number of optimisation steps reached')
+                return True
+
+            if 'fatal error' in line or 'input file has failed' in line:
+                logger.error('Fatal error in QChem calculation. Final lines:')
+                calc.print_final_output_lines(n=50)
+                return False
+
+        return True if calc_started else False
+
+    def optimisation_converged(self, calc) -> bool:
+        return any('OPTIMIZATION CONVERGED' in line
+                   for line in calc.output.file_lines)
+
+    def optimisation_nearly_converged(self, calc) -> bool:
+        # TODO: something better here
+        return False
+
+    def get_final_atoms(self, calc) -> List[Atom]:
+
+        if isinstance(calc.input.keywords, kws.HessianKeywords):
+            logger.warning('Hessian calculation performed - no change to '
+                           'geometry')
+            return calc.molecule.atoms
+
+        atoms = []
+
+        for i, line in enumerate(calc.output.file_lines):
+
+            if 'Coordinates (Angstroms)' not in line:
+                continue
+
+            """e.g.
+            
+                               Coordinates (Angstroms)
+             ATOM                X               Y               Z
+              1  O         0.0003489977   -0.1403224128    0.0000000000
+              2  H        -0.7524338562    0.4527672831    0.0000000000
+              3  H         0.7551329498    0.4500625364    0.0000000000
+            Point Group: cs    Number of degrees of freedom:     3
+            """
+
+            start_idx = i+2
+            end_idx = start_idx + calc.molecule.n_atoms
+
+            atoms = [Atom(_l.split()[1], *_l.split()[2:5])
```

Comment:
Yep exactly – needs the final set of coordinates. It goes forward because (1) it's more readable to not have two reversed() calls (atoms need to be in the same order) and (2) it's more tricky to pass the file in reverse as there might not be a consistent end line (e.g. if the calculation has been killed). It's such a fast function (c.f running DFT) that I'm inclined to leave it as is. 

_Premature optimization is the root of all evil_